src/factom-event-emitter.js

```diff
@@ -0,0 +1,367 @@
+     * emit their own events, and those events may have their own listeners. Therefore, this function also
+     * accepts an event argument, which is used to check whether there are any listeners still dependent on
+     * those emitted events. If there are, it does not remove the _fetch listener from the directoryBlock event.
+     * @private
+     */
+    _removeDirectoryBlockListener(listener, event) {
+        const hasListeners = this.listenerCount(event);
+        if (!hasListeners) {
+            return this.removeListener(this.event.directoryBlock, listener);
+        }
+    }
+
+    /**
+     * Listener that triggers on a 'directoryBlock' event. It fetches then emits the factoid block referenced
+     * in the emitted directory block.
+     * @param {DirectoryBlock} directoryBlock
+     * @private
+     */
+    async _fetchFactoidBlock(directoryBlock) {
+        try {
+            const factoidBlock = await this._cli.getFactoidBlock(directoryBlock.factoidBlockRef);
+            this.emit(this.event.factoidBlock, factoidBlock);
+        } catch (err) {
+            this.emit('error', err);
+        }
+    }
+
+    /**
+     * Listener that triggers on a 'directoryBlock' event. It fetches then emits the admin block referenced
+     * in the emitted directory block.
+     * @param {DirectoryBlock} directoryBlock
+     * @private
+     */
+    async _fetchAdminBlock(directoryBlock) {
+        try {
+            const adminBlock = await this._cli.getAdminBlock(directoryBlock.adminBlockRef);
+            this.emit(this.event.adminBlock, adminBlock);
+        } catch (err) {
+            this.emit('error', err);
+        }
+    }
+
+    /**
+     * Listener that triggers on a 'directoryBlock' event. It fetches then emits the entry credit block referenced
+     * in the emitted directory block.
+     * @param {DirectoryBlock} directoryBlock
+     * @private
+     * @async
+     */
+    async _fetchEntryCreditBlock(directoryBlock) {
+        try {
+            const entryCreditBlock = await this._cli.getEntryCreditBlock(directoryBlock.entryCreditBlockRef);
+            this.emit(this.event.entryCreditBlock, entryCreditBlock);
+        } catch (err) {
+            this.emit('error', err);
+        }
+    }
+
+    /**
+     * Listener that triggers on a 'directoryBlock' event. It iterates through entry blocks referenced in the directory block
+     * to find new chains. When a new chain is found, it emits the referenced entry block.
+     */
+    async _fetchNewEntryChains(directoryBlock) {
+        try {
+            const checkIfNewChainAndEmit = async ref => {
+                const entryBlock = await this._cli.getEntryBlock(ref.keyMR);
+
+                if (entryBlock.sequenceNumber === 0) {
+                    this.emit(this.event.newEntryChain, entryBlock);
+                }
+            };
+            await Promise.map(directoryBlock.entryBlockRefs, checkIfNewChainAndEmit, { concurrency: 5 });
```

Comment:
Oh nice, I didn't know about this concurrency option! I implemented something convoluted to achieve similar result... https://github.com/PaulBernier/factomjs/blob/master/src/add.js#L170
Nice find, thanks!